lib/AstOfLlbc.ml

```diff
@@ -566,47 +680,190 @@ let maybe_addrof (env: env) (ty: C.ty) (e: K.expr) =
+  "core::iter::range::Step";
+  (* TODO: for now, we leave into as-is in the AST, do a later pass that eliminates all identity
+     calls to into (post-monomorphization), and error our if there are any left that do not operate
+     on primitive types. We should probably remove the special-case in this file and treat it
+     generically with a dedicated pass over the krml ast. *)
+  "core::convert::From";
+  (* TODO: figure out what to do with those *)
+  "core::clone::Clone";
+  "core::fmt::Debug";
+]
+
+(* Using tests/where_clauses_simple as an example.
+
+ fn double<T: Ops + Copy, U: Ops+Copy> (...)
+
+ this gets desugared to fn double<T,U> where
+   T: Ops,      <-- ClauseId 0 (required_methods: add, of_u32)
+   T: Copy,     <-- ClauseId 1 (builtin, so neither required nor provided methods)
+   U: Ops,      <-- ClauseId 2 (required_methods: add, of_u32)
+   U: Copy,     <-- ClauseId 3 (builtin, so neither required nor provided methods)
+
+ the types we obtain by looking up the trait declaration have Self as 0
+ (DeBruijn).
+ *)
+let build_trait_clause_mapping env (trait_clauses: C.trait_clause list) =
+  List.concat_map (fun tc ->
+    let { C.clause_id; trait_id; clause_generics; _ } = tc in
+    let trait_decl = env.get_nth_trait_decl trait_id in
+
+    let name = string_of_name env trait_decl.name in
+    if List.mem name blocklisted_trait_decls then
+      []
+
+    else begin
+      (* FYI, some clauses like Copy have neither required nor provided methods. *)
+      Krml.KPrint.bprintf "clause decl %s id %d: FYI, clause_generic type is %s (required: %d) (provided: %d)\n"
+        name
+        (C.TraitClauseId.to_int clause_id)
+        (String.concat " ++ " (List.map C.show_ty (clause_generics.C.types)))
+        (List.length trait_decl.C.required_methods)
+        (List.length trait_decl.C.provided_methods);
+
+      (* The type variable from the function that the trait clause is applied to *)
+      let tvarid =
+        match clause_generics.C.types with
+        | [ TVar tvarid ] -> tvarid
+        | _ -> failwith "Clause above does not refer to a type var? confused"
+      in
+
+      List.map (fun (item_name, decl_id) ->
+        let decl = env.get_nth_function decl_id in
+        (clause_id, item_name), (tvarid, trait_decl.C.name, decl.C.signature)
+        ) trait_decl.C.required_methods @
+      List.map (fun (item_name, decl_id) ->
+        match decl_id with
+        | Some decl_id ->
+            let decl = env.get_nth_function decl_id in
+            (clause_id, item_name), (tvarid, trait_decl.C.name, decl.C.signature)
+        | None ->
+            failwith ("TODO: handle provided trait methods, like " ^ item_name)
+            ) trait_decl.C.provided_methods
+    end
+  ) trait_clauses
+
+(* Interpret a Rust function type, with trait bounds, into the krml Ast, providing:
+   - the number of polymorphic type variables
+   - the cg types, which only contains the original Rust const generic variables (not trait methods)
```

Comment:
What do you mean by "not trait methods"?

Reply:
Removed this. It's clear from the following lines that binders for the dictionary-style passing of trait methods are in arg_types.